HIF API documentation

MathisMARION opened the pull request:
I recommend reviewing the whole PR instead of individual commits.

The markdown view can be seen [outside of the PR](https://github.com/MathisMARION/wisun-br-linux/blob/wip/mm/rcpv2-doc/HIF.md). Bookmarks are available if you click the top right icon in the markdown view.

MathisMARION committed:
doc: update HIF frame structure

The frame payload description has been deported to its own section, to
clarify that it applies to both native UART and CPC.

Now the 'len' field only describes the frame payload. This way 0 is a
valid value, while before 'len' had to count the mandatory 'fcs' field.

The 'cmd' field has been changed to a byte. If the command space needs
to be expanded in the future, an ID for extended commands will be
allocated.

WI_SUN-2436

MathisMARION committed:
doc: drop mention of LUTT fill

This IE is only inserted by LFNs.

MathisMARION committed:
doc: introduce SET RCP command prefix

MathisMARION committed:
doc: shorten some RCP command names

MathisMARION committed:
doc: remove unimplemented commands

MathisMARION committed:
doc: editorial changes and typos

Co-authored-by: Kylian Balan <kylian.balan@silabs.com>
Co-authored-by: Hadrien Hoyer <hadrien.hoyer@silabs.com>
Co-authored-by: Thibaut Colin <thibaut.colin@silabs.com>
Co-authored-by: Ayoub Aba Haddou <ayoub.abahaddou@silabs.com>

MathisMARION committed:
Revert "doc: remove HIF.md"

This reverts commit 0e61412c1b04035e9d92fb9427fad5450c661ef1.

MathisMARION committed:
doc: update HIF REQ_DATA_TX_ABORT

MathisMARION committed:
doc: remove HIF REQ_DATA_TX_ABORT

This command is currently so buggy that it is basically unusable. Do
not document it until it works properly.

MathisMARION committed:
doc: update HIF *FILTER* commands

MathisMARION committed:
doc: update HIF *SEC* commands

MathisMARION committed:
doc: properly define subsections titles in HIF.md

MathisMARION committed:
doc: add table for HIF FHSS commands

MathisMARION committed:
doc: update HIF *FHSS* commands

MathisMARION committed:
doc: update HIF IND_DATA_RX

MathisMARION committed:
doc: update general observation

jerome-pouiller commented on HIF.md:
```diff
@@ -34,21 +34,25 @@ structure is defined by the CPC specification.
+With both protocols, the frame payload is in the following form:
+
+ - `uint8_t cmd`  
+   Command number.
```
Indentation

jerome-pouiller commented on HIF.md:
```diff
@@ -34,21 +34,25 @@ structure is defined by the CPC specification.
+   Command number.
+
+ - `uint8_t body[]`  
+   Command body.
```
Ditto

jerome-pouiller commented on HIF.md:
```diff
@@ -448,6 +450,30 @@ FIXME: define default value
+| Unicast to LFN            | N/A                     | [`REQ_DATA_TX`][tx-req]     |
+| Unicast PAN Advert to LFN | N/A                     | [`REQ_DATA_TX`][tx-req]     |
+| Broadcast to LFN          | N/A                     | [`SET_FHSS_LFN_BC`][bc-lfn] |
+| Asynchronous              | N/A                     | [`SET_FHSS_ASYNC`][async]   |
```
- Maybe it makes sense to repeat "location of parameter" in the columns headers.
 - Does "Reception" of "Unicast to FFN" makes sense?
 
Maybe:
```
|  FHSS type                   | Location of FHSS parameters |
|------------------------------|-----------------------------|
| Rx Unicast                   | [`SET_FHSS_UC`][uc]         |
| Rx Broadcast                 | [`SET_FHSS_FFN_BC`][bc]     |
| Tx Unicast to FFN            | [`REQ_DATA_TX`][tx-req]     |
| Tx Broadcast to FFN          | [`REQ_DATA_TX`][tx-req] and [`SET_FHSS_FFN_BC`][bc] |
| Tx Unicast to LFN            | [`REQ_DATA_TX`][tx-req]     |
| Tx Unicast PAN Advert to LFN | [`REQ_DATA_TX`][tx-req]     |
| Tx Broadcast to LFN          | [`SET_FHSS_LFN_BC`][bc-lfn] |
| Tx Asynchronous              | [`SET_FHSS_ASYNC`][async]   |
```

jerome-pouiller commented on HIF.md:
```diff
@@ -474,154 +474,108 @@ of these commands for a more detailed explanation.
-  TBD
+  - `uint8_t chan_mask[]`  
+     Bitmask of channels to use for transmission. See the same field in
+     ["Channel sequence"][chan-seq] for more details.
```
`SET_FHSS_ASYNC` do not use a `struct chan_seq`?

MathisMARION replied:
No, there's no such thing as a channel function for async.

jerome-pouiller commented on HIF.md:
```diff
@@ -601,41 +601,41 @@ Only present if KEY_IDENTIFY_MODE_INDEX:
-If `FILTER_BAD_SRC64` is set, drop frames coming from denied extended addresses.
+Refuse frames based on the source address. This should only be used in specific
+testing scenarios to force network topologies. By default, the address list is
+empty.
```
I believe there are frame with unknown source address in the 15.4 standard (at least they exist in 802.11). I assume they are are not filtered?

MathisMARION replied:
These frames are currently always refused by the RCP frame parser (they are not used in Wi-SUN). I should probably add some more details as to what is accepted in `REQ_DATA_TX` and `IND_DATA_RX`, similar to the short paragraph about the auxiliary security header.

jerome-pouiller commented on HIF.md:
```diff
@@ -385,69 +386,97 @@ Only present if `HAVE_MODE_SWITCH_STAT`:
+    Enable mode switch in RX and TX, using all the PHY configurations available
+    in the same group as the selected PHY. Mode switch is enabled at PHY level
+    (PHR) and MAC level (MAC command frame). See [`REQ_DATA_TX`][tx-req] for TX
+    parameters.
```
I thought this field only impacts Rx (Tx is managed by `REQ_DATA_TX`)

MathisMARION replied:
I need to check with Alban. Code makes it not obvious.

jerome-pouiller commented on HIF.md:
```diff
@@ -385,69 +386,97 @@ Only present if `HAVE_MODE_SWITCH_STAT`:
-### `0x25 SET_RADIO_TX_POWER`
+    - `0`: None (disabled, default)
+    - `2`: Japan ([ARIB][arib])
+    - `5`: India ([WPC][wpc])  
```
Maybe we can mention the reserved values:
```
 - `uint32_t value`  
    - `0`: None (disabled, default)
    - `1`: Reserved (future restrictions for North America (FCC))
    - `2`: Japan ([ARIB][arib])
    - `3`: Reserved (future restrictions for Europe (ETSI))
    - `4`: Reserved
    - `5`: India ([WPC][wpc])

MathisMARION replied:
What is the benefit?

Also I wanted to give a bit more details regarding what behavior changes, but I realized that I do not know what ARIB does, I'll take the time to talk to fill that info.

Kilio22 replied:
I don't think this specification is the place where an explanation should be written, is it?
To my eyes, links to the various specifications is enough.

It might also be smart from a customer point of view to not give any details here, they will reach out to us if they want to know more.

MathisMARION replied:
The thing is, we are most definitely not implementing all features requested by the regulation specs, we do not want to pretend that these settings ensure a regional certification.

Kilio22 replied:
I am now asking myself "what are we marketing" about that?
Let's be honest, I believe we are not entirely transparent on that anyways.

This should be documented on DSC.

silabs-ThibautC replied:
I agree with Kilio22, I think we'll probably end up with something on dsc. We'll point to that page then.

MathisMARION replied:
atm there are no details given with the doc for [`sl_wisun_set_regulation()`](https://docs.silabs.com/wisun/2.2.0/wisun-stack-api/sl-wisun-api#sl-wisun-set-regulation).

jerome-pouiller commented on HIF.md:
```diff
@@ -164,141 +164,113 @@ This command should be sent after `IND_RESET`, before any other command.
-      - If the auxiliary security header is present, the device will encrypt the
-        15.4 payload (see `SET_SEC_KEY`) and fill the MIC and the frame
-        counter
+      - Timing parameters in Wi-SUN IEs (see ["FHSS Configuration"][fhss])
```
... IEs (UTT, BT, LBT) (see ...

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+## Frame structure
+
+The RCP can use the _Native UART_ or the _CPC_ protocol. With CPC, the frame
+structure is defined by the CPC specification.
```
Link to CPC?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+ - `uint16_t fcs`  
+    [CRC-A][fcs] of the `payload` field.
+
+With both protocols, the frame payload is in the following form:
```
`Regardless of the protocol used, the frame payload always has the following structure:`

Feel free to disregard my comments (or not) when they are subjective like that.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+No-operation. Can be used to synchronize UART communication.
+
+ - `uint8_t garbage[]`  
+    Extra data may be included. Must be ignored.
```
Either `Extra data may be included. Must be ignored.`, either `Data after command is ignored.`, no?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+    If set, start the bootloader instead of normal firmware. Set this field if
+    you want to flash a new firmware. Refer to the bootloader documentation to
+    know how to send the new firmware (usually, the XMODEM protocol is used to
+    send the new firmware other the UART).
```
I don't think `other` is the right word here.

Maybe stopping at `new firmware` and linking the documentation is better?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+Commands don't reply in case of success. If the device gets misconfigured or if
+any other fatal error happens, the device will reset. In this case, `IND_FATAL`
+may be sent before `IND_RESET`. It contains information about the error. It can
+be used display the error to the user for debugging purpose.
```
`to display`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+
+### `0x06 SET_HOST_API`
+
+Inform RCP about the host API version. If not send, the RCP will assume `2.0.0`.
```
`the RCP`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+      - Frame counter (see ["Security"][sec])
+      - Message Integrity Code (MIC) (see ["Security"][sec])
+
+    <!-- TODO: Explain EDFE -->
```
Are we leaving this todo?

MathisMARION replied:
I am not willing to document EDFE for now since we do not want it do be used by customers.

hhoyer replied:
Then it probably shouldn't be mentioned at all

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+            `FHSS_DEFAULT == 0` only.
+         - `0x03 FHSS_TYPE_LFN_BC`: Broadcast to LFN  
+            `FHSS_DEFAULT == 1` only, see [`SET_FHSS_LFN_BC`][bc-lfn].
+         - `0x04 FHSS_TYPE_ASYNC`:  Asynchronous (MLME-WS-ASYNC-FRAME)  
```
Is it relevant to mention `MLME-WS-ASYNC-FRAME`?

MathisMARION replied:
I think it's worth mentioning once in our doc, but maybe it should be in the FHSS section instead.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+    will use the `timestamp_us` value from [`CNF_DATA_TX`][tx-cnf] and
+    [`IND_DATA_RX`][rx].
+ - `uint24_t ufsi` (from UTT-IE)
+ - `uint8_t dwell_interval` (from US-IE)
```
Space missing (for alignment)

MathisMARION replied:
Note that consecutive spaces are always rendered as 1 space, even in monospace formatting (at least with the VSCode renderer).

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+ - `uint8_t  slot_duration_ms` (from LND-IE)
+ - `uint8_t  slot_count` (from LND-IE)
+ - `uint16_t slot_first` (from LND-IE)
+
```
Should we mention anything regarding the other FHSS types?
Like `Other FHSS types have no additional fields.`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+Only present if `FHSS_DEFAULT == 0`:
+
+ - `struct chan_seq`  
+   [Channel sequence][chan-seq] description.
```
`See ...`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+
+ - `struct rate_config[4]`  
+    Fixed length and ordered array of rates to attempt until the transmission is
+    succesful, or all rates been tried.
```
`successful`

Kilio22 replied:
`have been tried`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+### `0x12 CNF_DATA_TX`
+
+Status of an earlier data request ([`REQ_DATA_TX`][tx-req]). Also returns the
+acknowledge frame if relevant.
```
`the IEEE 802.15.4 acknowledge frame...`

Kilio22 replied:
`acknowledgment` (all words are missing the `d` ahah)

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+    Length of the IEEE 802.15.4 acknowlegment frame if received.
+
+ - `uint8_t frame[]`  
+    Unencrypted IEEE 802.15.4 acknowlegment frame (see ["Security"][sec]).
```
`Decrypted`?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+
+ - `uint8_t frame[]`  
+    Unencrypted IEEE 802.15.4 acknowlegment frame (see ["Security"][sec]).
+    <!-- TODO: Explain EDFE. -->
```
Are we leaving this TODO?

silabs-AyoubA replied:
What's the plan here?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+    <!-- TODO: Explain EDFE. -->
+
+ - `uint64_t timestamp_us`  
+    The timestamp (relative to the date of the RCP reset) when the frame have
```
I don't think mentioning `relative to ...` is necessary considering you already mentioned it at the beginning.

Kilio22 replied:
`when the frame has been received`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+    Frame counter used in the successful transmission (see `frame_counter` in
+    [`SEC_SET_KEY`][key]). Only valid if the request included an auxiliary
+    security header.
+    <!-- TODO: Return the last used frame counter, even if unsuccessful. -->
```
Todo?

MathisMARION replied:
This is more a TODO in RCP code. The issue currently is that the frame counter is incremented for all retries, but it is not returned to the host in case of TX failure.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+
+ - `uint8_t tx_failures`  
+    Number of transmission failures (without counting CCA failures).
+    <!-- TODO: Make this description more precise. -->
```
Todo?

MathisMARION replied:
This is an absolute pain to describe. I am not sure if a CSMA-CA failure counts as a TX failure or if it's only for acknowledge frames that did not receive an ACK.

In general, we probably want a clear description of how the RCP handles retries and CSMA-CA. We will need Alban's help for that.

Kilio22 replied:
Have you tried Copilot? ahah

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+    <!-- TODO: Return the last used frame counter, even if unsuccessful. -->
+
+ - `uint16_t chan_num`  
+    Channel number used in the succesful transmission.
```
`successful`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+    Channel number used in the succesful transmission.
+
+ - `uint8_t cca_failures`  
+    Number of times channel access has failed in the last CSMA-CA attempt.
```
I'm not sure if `the last CSMA-CA attempt` is precise enough 🤔

MathisMARION replied:
Again, this is stupid to describe. I am not even sure why we decided to include this metric.
The CCA failure counter is reset when a CSMA-CA is started, which supposedly happens at every retry.
So only the last retry is counted from what I understood from the code.

Kilio22 replied:
`Since the last TX attempt`? 🤔

Kilio22 replied:
Btw, should we document somewhere the maximum number of retries?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+
+| Value | Description
+|-------|-------------
+|`0x00` | Success.
+|`0x01` | Not enough memory available.
+|`0x02` | CCA failure on all channel access attempts.
+|`0x03` | No ACK received (if ACK bit set in request).
+|`0x04` | Frame spent too long in RCP (10s for unicast, 20s for broadcast, 40s for async, 300s for LFN).
+|`0x05` | RCP internal error (reach out Silicon Labs support).
```
Do you want to add a `Name` column here as well?

MathisMARION replied:
Does not seem necessary to me.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+
+### `0x13 IND_DATA_RX`
+
+Receive IEEE 802.15.4 frames. The RCP will start emitting these indications
```
`Reception of ...`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+    Length of the received IEEE 802.15.4 frame.
+
+ - `uint8_t frame[]`  
+    Unencrypted IEEE 802.15.4 frame, (see ["Security"][sec]).
```
`Decrypted`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+    Unencrypted IEEE 802.15.4 frame, (see ["Security"][sec]).
+
+ - `uint64_t timestamp_rx_us`  
+    The timestamp (relative to the date of the RCP reset) when the frame have
```
`has been received`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+Several schemes are used in Wi-SUN to transmit or receive packets using
+frequency hopping. These are independently defined using a set of parameters,
+and may differ between transmission and reception. The following table
+summarizes which commands configure the various FHSS parameters. Note that
```
If we are talking about pure configuration, only the `SET_*` commands should be documented here.
Maybe writing about which `SET` cmd should be called to **enable** the various RX/TX capabilities would be better?

Then we can have a separate section about Wi-SUN-specific timings filled up by the RCP on `REQ_DATA_TX` (like we already have, but I think a new section/subsection is necessary).

MathisMARION replied:
The goal of this section is to summarize all the schemes available for RX/TX and all that is necessary to make them work with the whole HIF API, not just the `SET_FHSS` commands. This paragraph should be referenced in the same way that the security section is referenced throughout the data TX/RX section (on that note maybe I'm missing a link somewhere in `REQ_DATA_TX`).

Wi-SUN is arguably quite convoluted when it comes to this so I think it is sane putting everything in a big picture.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+ - `struct chan_seq`  
+   [Channel sequence][chan-seq] description.
+
+<!-- TODO: document parent BS-IE following -->
```
Todo?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+Configure asynchronous transmissions for network discovery.
+
+ - `uint32_t tx_duration`  
+    Maximum number of milliseconds the RCP is allowed to stay in continuous TX
+    for an async transmission. If that duration is exceeded, the async
+    transmission is split into chunks of that duration until all channels have
+    been used. By default, the value is set to `0xffffffff`, which disables this
+    feature.
```
Should we write about the benefits of doing so?
something along the lines of `allowing RX and TX between chunks`?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+MAC layer data appears in cleartext in HIF commands so integrators are expected
+to secure the serial link, for example using [CPC with link encpryption](cpc).
+
+To encrypt transmitted IEEE 802.15.4 frames, an auxiliary security security
```
The word `security` is written twice

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+To encrypt transmitted IEEE 802.15.4 frames, an auxiliary security security
+header must be included in the buffer passed to [`REQ_DATA_TX`](#0x10-req_data_tx),
+and space must be reserved at the end of the frame for the Message Integrity
+Code (MIC) depending on the security level. The packet payload is not yet
```
Should we document the MIC size to use? Or at least link the relevant specification and section?

MathisMARION replied:
Users need to open the 15.4 spec to know how to craft a frame anyway, I would not point to any specific section.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+> responsability of the host to maintain frame counters per key and per
+> neighbor, and check them to prevent any replay attacks.
+
+Acknowledgement frames are unencrypted and sent to the host in the same manner
```
`decrypted`

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+
+### `0x5A SET_FILTER_SRC64`
+
+Refuse frames based on the source address. This should only be used in specific
```
`based on the MAC source address`? `based on the source address in the IEEE 802.15.4 header`?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+ - `uint8_t payload[]`  
+    Arbitrary data.
+
+## Simulation
+
+### `0xF0 IND_REPLAY_TIMER`
+
+Only for simulation purpose. The device will never send it.
+
+### `0xF1 IND_REPLAY_SOCKET`
+
+Only for simulation purpose. The device will never send it.
+
```
Should we even document it?

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,739 @@
+ - `uint32_t fw_version`  
+    Device version
+
+ - `char fw_version_str[]`  
```
Should we mention that the string is null terminated?

MathisMARION committed:
doc: mention CRC algorithm

MathisMARION committed:
doc: update HIF REQ_DATA_TX

MathisMARION committed:
doc: update HIF CNF_DATA_TX

MathisMARION committed:
doc: remove HIF IND_REPLAY*

These commands are not really part of the HIF API since the RCP cannot
send them. Documenting them here does not seem valuable.

MathisMARION commented:
v2:
- Fix typos and wording.
- Simplify FHSS table.
- Remove `IND_REPLAY` section.

MathisMARION commented:
One thing you may have not noticed by not looking at commits is that I removed documentation for `REQ_DATA_TX_ABORT` because the current implementation seems basically broken.

Kilio22 commented:
> One thing you may have not noticed by not looking at commits is that I removed documentation for `REQ_DATA_TX_ABORT` because the current implementation seems basically broken.

Maybe it's time to up this ticket?

MathisMARION committed:
doc: parent follow in HIF SET_FHSS_FFN_BC

MathisMARION commented:
v3:
- Document parent follow in `SET_FHSS_FFN_BC`.
- Minor rewording in `SET_FHSS_ASYNC`.

MathisMARION committed:
doc: detail HIF 15.4 packet format restriction

MathisMARION commented:
v4:
- Update mode switch enable description and add note on POM-IE.
- Describe CCA and retries.
- Describe packet format restrictions.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+but the way attempts are performed differs depending on the [FHSS type][fhss]
+used.
+
+Unicast and broadcast to FFNs use the CSMA-CA algorithm with the following
```
The `IEEE802.15.4 CSMA-CA algorithm`

Kilio22 replied:
I would bold `FFNs` to emphasize the difference.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+corresponds to the next available TX slot. This detail matters for counting
+CCA failures versus transmission failures.
+
+Asynchronous transmissions use a CCA per channel, and skip the channel on
+failure. No overall CCA failure can be returned in [`CNF_DATA_TX`][tx-cnf]
+with these transmissions.
```
Do we have to explain what `Asynchronous transmissions` are?

To be clear, I would insist the CSMA backoff is not used in asynchronous transmissions.
I would also say that `no retry is performed on failure` instead of `skip the channel on failure.`

MathisMARION replied:
> Do we have to explain what `Asynchronous transmissions` are?

I don't think so, for the same reason that we should not explain CSMA-CA in details but instead only provide the parameters.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+| `macMaxCsmaBackoffs`  | `8`   | Number of CSMA-CA backoffs to attempt before declaring failure.
+| `macMaxFrameRetries`  | `19`  | Number of retries allowed after a transmission failure.
+
+Unicast and broadcast transmissions to LFNs can be conceived as if they use
```
I would bold `LFNs` to emphasize the difference.

Kilio22 commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+corresponds to the next available TX slot. This detail matters for counting
+CCA failures versus transmission failures.
+
+Asynchronous transmissions use a CCA per channel, and skip the channel on
```
I would bold `Asynchronous transmissions`

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+# RCP Hardware Interface
+
+This document describes the Hardware InterFace (HIF). HIF is the API between
+`wsbrd` and the RCP device.
```
I would not mention wsbrd here. It is the RCP API. One of the reasons why we want to document it exhaustively is to promote its use in other projects.

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+  - `SET`: special case of `REQ` to configure the device
+  - `CNF` (confirmation): a reply from the device to the host
+  - `IND` (indication): a spontaneous frame from the device to the host
+
```
Do we want to make it explicit that requests will not receive a confirmation by default?

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+Commands don't reply in case of success. If the device gets misconfigured or if
+any other fatal error happens, the device will reset. In this case, `IND_FATAL`
+may be sent before `IND_RESET`. It contains information about the error. It can
+be used to display the error to the user for debugging purpose.
```
purposes

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+[tx-cnf]: #0x12-cnf_data_tx
+[rx]:     #0x13-ind_data_rx
+
+### Channel Access and Retries
```
Is it essential for the comprehension of the HIF API documentation?

MathisMARION replied:
I think this is necessary to explain `cca_failures` and `tx_failures` in `CNF_DATA_TX`.
Also some users might be interested in the CSMA-CA parameters in use, although I am aware that documenting this may raise unwanted questions from customers.

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+
+ - `struct rate_config[4]`  
+    Fixed length and ordered array of rates to attempt until the transmission is
+    successful, or all rates have been tried. Mode switch only makes sense with
```
`Mode switch only makes sense with unicast transmissions.`

I'd either remove that sentence or move it to the flag documentation. 

Did you mean that it was not supported?

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+### `0x12 CNF_DATA_TX`
+
+Status of an earlier data request ([`REQ_DATA_TX`][tx-req]). Also returns the
+IEEE 802.15.4 acknowledgement frame if relevant.
```
if relevant = if the tx was a unicast?

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+    frame is transmitted on the radio but no acknowledgement is received.
+
+ - `uint8_t tx_failures`  
+    Number of transmission failures (without counting CCA failures).
```
"does not account for CCA failures" may sound better?

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+|`0x01` | Not enough memory available.
+|`0x02` | Channel access failure (CCA failure or busy RX).
+|`0x03` | No ACK received (if ACK bit set in request).
+|`0x04` | Frame spent too long in RCP (10s for unicast, 20s for broadcast, 40s for async, 300s for LFN).
```
I'm curious, what could cause that timeout? Is it a failsafe in case a packet in stuck in the RCP?

MathisMARION replied:
I see this as a failsafe. I believe it can be triggered if there are many packets in the RCP queues.

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+        - `0x0001`: If set, this entry is in same group than the previous.
+        - `0x01FE`: Bitfield of supported OFDM MCS (from MCS0 to MCS11).
+    - `uint8_t rail_phy_mode_id`  
+       Wi-SUN PhyModeId, but for OFDM, only MCS0 exists.
```
That isn't very clear. 

For OFDM, only set the PhyType. 
For OFDM, only set the Option number.

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+ - `bool enable_ms` (API > 2.0.1)  
+    Enable mode switch in reception at PHY level (PHR), using all the PHY
+    configurations available in the same group as the selected PHY. Note that
+    mode switch  cannot be disabled at MAC level (MAC command frame).
```
I see two white spaces between switch and cannot.

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+ - `uint8_t mcs`  
+    MCS to be used if `index` points to an OFDM modulation.
+
+ - `bool enable_ms` (API > 2.0.1)  
```
FYI: I'm not saying it should be changed but I read "enable milliseconds".

silabs-ThibautC commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+
+The RCP is responsible for encrypting and decrypting IEEE 802.15.4 frames.
+MAC layer data appears in cleartext in HIF commands so integrators are expected
+to secure the serial link, for example using [CPC with link encpryption][cpc-sec].
```
I think we can insist even more here. We could clarify that the system will be exposed to packet injection and eavesdropping if it's not done.

hhoyer commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+The type `bool` is a `uint8_t`, but only the LSB is interpreted. All the other
+bits are reserved.
+
+When a bit is reserved or unassigned, it be set to 0 by the transmitter and must
```
missing "must": it _must_ be set to 0

hhoyer commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+
+ - `uint16_t len`  
+    Length of `payload`. Only the 11 least significant bits (`0x07FF`) are
+    used. The 5 most significant bits (`0xF800`) must ignored.
```
must _be_ ignored

hhoyer commented on HIF.md:
```diff
@@ -0,0 +1,810 @@
+
+> [!WARNING]
+> The current RCP implementation uses the same set of parameters for FFN and
+> LFN brodcast channel sequences. Thus, [`SET_FHSS_FFN_BC`][bc] and
```
typo _brodcast_

MathisMARION committed:
doc: update HIF *RADIO* commands

MathisMARION commented:
v5:
- Emphasize bus security concerns.
- Typos and minor editorial changes.

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+# RCP Hardware Interface
+
+This document describes the Hardware InterFace (HIF) for the Silicon Labs
```
```suggestion
This document describes the Hardware Interface (HIF) for the Silicon Labs
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+## General observations
+
+In this specification, the prefix used in the command means:
+  - `REQ` (request): a command sent by host to the device
```
```suggestion
  - `REQ` (request): a command sent by the host to the device
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+The type `bool` is a `uint8_t`, but only the LSB is interpreted. All the other
+bits are reserved.
+
+When a bit is reserved or unassigned, it be must set to 0 by the transmitter
```
```suggestion
When a bit is reserved or unassigned, it must be set to 0 by the transmitter
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+  - `0x00FFFF00` Minor revision
+  - `0x000000FF` Patch
+
+All timestamps used are in the RCP timebase and in microseconds, with origin
```
Just a reformulation for better clarity
```suggestion
All timestamps are based on the RCP timebase, measured in microseconds,
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+  - `0x000000FF` Patch
+
+All timestamps used are in the RCP timebase and in microseconds, with origin
+the device reset.
```
```suggestion
with the origin set at the device reset.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+the device reset.
+
+All channel numbers are defined as if there were no channel mask applied
+(_ChanN = Chan0 + n * ChanSpacing_).
```
Is N in ChanN the same as the n in the formula, if it is the case we can write change the formula for clarity:
```suggestion
(_Chan_n = Chan0 + n * ChanSpacing_).
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+error detection.
+
+ - `uint16_t len`  
+    Length of `payload`. Only the 11 least significant bits (`0x07FF`) are
```
```suggestion
    Length of the `payload`. Only the 11 least significant bits (`0x07FF`) are
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    If set, start the bootloader instead of normal firmware. Set this field if
+    you want to flash a new firmware. Refer to the bootloader documentation to
+    know how to send the new firmware (usually, the XMODEM protocol is used to
+    send the new firmware over the UART).
```
```suggestion
    If set to `true`, the bootloader will start instead of the normal firmware. Use this 
    option when flashing new firmware. For details on how to upload the new 
    firmware (typically using the XMODEM protocol over UART), refer to the 
    bootloader documentation.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+### `0x01 REQ_NOP`
+
+No-operation. Can be used to synchronize UART communication.
+
```
I suggest adding this header to all the commands

```suggestion

#### Command Body

```

MathisMARION replied:
Nah it generates a unnecessary bookmark and does not bring much more clarity imo.

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    EUI64 flashed in the device
+
+ - `uint8_t reserved[]`  
+    Extra data may be included. For backward compatibility, they must be
```
```suggestion
    Extra data may be included. For backward compatibility, it must be
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    `1.5.0-5-ga91c352~bpo`).
+
+ - `uint8_t hw_eui64[8]`  
+    EUI64 flashed in the device
```
```suggestion
    EUI64 flashed on the device
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+Commands don't reply in case of success. If the device gets misconfigured or if
+any other fatal error happens, the device will reset. In this case, `IND_FATAL`
+may be sent before `IND_RESET`. It contains information about the error. It can
+be used to display the error to the user for debugging purposes.
```
The IND_FATAL may be sent or will be sent before the `IND_RESET`?
```suggestion
Commands do not send a reply upon success. However, if the device becomes
misconfigured or encounters a fatal error, it will reset. In such cases, an
`IND_FATAL` message may be sent before `IND_RESET`, providing details about 
the error. This information can be used to display the error to the user for 
debugging purposes.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+be used to display the error to the user for debugging purposes.
+
+ - `uint16_t error_code`  
+    See the table below for the list of errors.
```
```suggestion
    Refer to the table below for the detailed list of errors.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    See the table below for the list of errors.
+
+ - `char error_string[]`  
+    Human readable error.
```
```suggestion
    Human-readable error message.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+|`EINVAL_ADDR_MODE`    |`0x100e`| Invalid IEEE 802.15.4 address mode.                         |
+|`EINVAL_SCF`          |`0x100f`| Invalid IEEE 802.15.4 security control field.               |
+|`EINVAL_FRAME`        |`0x1010`| Invalid IEEE 802.15.4 frame (generic).                      |
+|`EINVAL_CHAN_FIXED`   |`0x1011`| Invalid fixed channel.                                      |
```
I prefer `EINVAL_FIXED_CHAN` :sweat_smile:
```suggestion
|`EINVAL_FIXED_CHAN`   |`0x1011`| Invalid fixed channel.                                      |
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+
+Inform the RCP about the host API version. If not send, the RCP will assume
+`2.0.0`. This command should be sent after `IND_RESET`, before any other
+command.
```
```suggestion
Informs the RCP of the host API version. If this command is not sent, the
RCP will assume the API version is `2.0.0`. This command must be sent
after `IND_RESET` and before any other command.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+Only a subset of the IEEE 802.15.4 frame formats are supported for the needs of
+Wi-SUN:
+
+  - Frame type must be `1` (data) or `2` (ack, [`CNF_DATA_TX`][tx-cnf] only).
```
```suggestion
  - Frame type must be either `1` (data) or `2` (ack, [`CNF_DATA_TX`][tx-cnf] only).
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+### Channel Access and Retries
+
+In Wi-SUN, all transmissions require the use of Clear Channel Assessment (CCA),
+but the way attempts are performed differs depending on the [FHSS type][fhss]
+used.
```
```suggestion
but the way attempts are performed differs depending on the used
[FHSS type][fhss].
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+
+***Asynchronous*** transmissions use one CCA per channel, and skip the channel on
+failure without attempting any retry. No overall CCA failure can be returned in
+[`CNF_DATA_TX`][tx-cnf] with these transmissions.
```
```suggestion
***Asynchronous*** transmissions use one CCA per channel and skip the channel on
failure without attempting any retries. No overall CCA failure is returned in
[`CNF_DATA_TX`][tx-cnf] for these transmissions.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+
+Send a IEEE 802.15.4 frame. All data requests are followed by a data
+confirmation ([`CNF_DATA_TX`][tx-cnf]) once the transaction has finished, even
+in case of transmission failure or abort.
```
```suggestion
Sends an IEEE 802.15.4 frame. All data requests are followed by a data
confirmation ([`CNF_DATA_TX`][tx-cnf]) once the transaction is complete, even
in case of transmission failure or abort.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+ - `uint8_t frame[]`  
+    A IEEE 802.15.4 frame without PHR and FCS. The RCP reads and interprets the
+    header to determine what processing needs to be done. The frame must not be
+    encrypted at this stage, see ["Security"][sec] for more details.
```
```suggestion
    An IEEE 802.15.4 frame without PHR and FCS. The RCP reads and interprets the
    header to determine what processing is needed to be done. The frame must not
     be encrypted at this stage, see ["Security"][sec] for more details.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    If the following fields are present, they are automatically filled by the
+    RCP (their input value is ignored and overwritten):
+      - Sequence number
+      - Timing parameters in Wi-SUN IEs (UTT, BT, LBT, LTO) (see
```
I think we should add an abbreviation section

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    is successful, or all rates have been tried.
+     - `uint8_t phy_mode_id`: Wi-SUN _PhyModeId_, must be in the group selected
+        with [`SET_RADIO`][rf-set].
+     - `uint8_t tx_attempts`: Maximum number of attempts for this rate, try the
+        next entry once this is exceeded. Note that this overrides
+        `macMaxFrameRetries` described in ["Channel Access and Retries"][cca].
```
```suggestion
     - `uint8_t tx_attempts`: The maximum number of attempts allowed for the
        selected rate. Once this limit is exceeded, the next entry will be tried.
        Note that this value overrides `macMaxFrameRetries` described in
        ["Channel Access and Retries"][cca].
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+     - `uint8_t tx_attempts`: Maximum number of attempts for this rate, try the
+        next entry once this is exceeded. Note that this overrides
+        `macMaxFrameRetries` described in ["Channel Access and Retries"][cca].
+     - `int8_t tx_power_dbm`: TX power to use with this entry, saturates to the
+        value configured with [`SET_RADIO_TX_POWER`][rf-pow].
```
```suggestion
     - `int8_t tx_power_dbm`: The TX power to be used with this entry, limited
        to the maximum value configured with [`SET_RADIO_TX_POWER`][rf-pow].
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    The timestamp (relative to the date of the RCP reset) when the frame has
+    been received.
+
+ - `uint8_t lqi`  
+    Received Link Quality Indicator (LQI) as reported by [RAIL][lqi] (ACK
+    received only).
```
we should drop this one

MathisMARION replied:
The API is already in use, it cannot be simply dropped unfortunately.

jerome-pouiller replied:
BTW, why don't we want to report it?

silabs-ThibautC replied:
Because it is not used and may be misleading. It's opening the door to questions like "why my parent is not that router instead, its lqi is better than the one chosen by RPL"

jerome-pouiller replied:
This API has been designed to allow customers to do customisation. This is the reason why we expose all the data we can rather than only the needed one.

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    The timestamp (relative to the date of the RCP reset) when the frame has
+    been received on the device.
+
+ - `uint8_t lqi`  
+    Received Link Quality Indicator (LQI) as reported by [RAIL][lqi].
```
Same here, we should drop LQI

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    Maximum transmission power in dBm. The RCP may utilize a lower value based
+    on internal decision making or hardware limitations but will never exceed
+    the given value. The default value is 14dBm.
+
```
```suggestion
 - `int8_t tx_power_dbm`  
    Maximum transmission power in dBm. The RCP may use a lower value based
    on internal decision-making or hardware limitations but will never exceed
    the given value. The default value is 14dBm.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+## Frequency Hopping (FHSS) configuration
+
+Several schemes are used in Wi-SUN to transmit or receive packets using
+frequency hopping. These are independently defined using a set of parameters,
```
```suggestion
frequency hopping. These are independently defined using a set of parameters
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+and may differ between transmission and reception. The following table
+summarizes which commands configure the various FHSS parameters. Note that
+[`REQ_DATA_TX`][tx-req] always needs to be called for transmission, but timing
+parameters are sometimes passed with the command, and sometimes stored in the
```
"sometimes" is vague can we say exactly when?

MathisMARION replied:
"sometimes" is described in the table, I do not think there is a simple way to write it down.

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+     Bitmask of used channels (1 for used, 0 for excluded). Channels are
+     counted byte by byte, from least to most significant bit (channel `n` maps
+     to `chan_mask[n / 8] & (1 << (n % 8))`). Integrators are responsible for
+     meeting local regulation contraints by excluding disallowed channels from
```
```suggestion
     meeting local regulation constraints by excluding disallowed channels from
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+    RCP timing from the host. By default, the address is set to
+    `ff:ff:ff:ff:ff:ff:ff:ff`, which disable parent following. Not including
+    this block in the command also disables parent following. The next
+    parameters set up the initial timings.
```
```suggestion
    MAC address of a neighboring node whose broadcast schedule should be
    followed (typically a RPL parent or an EAPoL target). The RCP will read
    the BT-IE of frames received from that source address to maintain its
    broadcast schedule synchronized, which prevents needing to regularly update
    RCP timing from the host. By default, the address is set to
    `ff:ff:ff:ff:ff:ff:ff:ff`, which disables the parent following. Not including
    this block in the command also disables the parent following. The next
    parameters set up the initial timings.
```

silabs-AyoubA commented on HIF.md:
```diff
@@ -0,0 +1,823 @@
+this command to stress and debug the link with the device.
+
+ - `uint16_t counter`  
+    Arbitrary value (usually an incrementing counter that will be send back in
```
```suggestion
    Arbitrary value (usually an incrementing counter that will be sent back in
```

MathisMARION committed:
doc: add description of CCA and retries in HIF.md

MathisMARION committed:
doc: update HIF IND_FATAL

MathisMARION committed:
doc: add acronyms section to HIF.md

jerome-pouiller commented on HIF.md:
```diff
@@ -5,6 +5,35 @@ Wi-SUN Radio Co-Processor (RCP). The RCP implements IEEE 802.15.4 frame
+| BT-IE   | Wi-SUN        | Broadcast Timing IE
+| FFN     | Wi-SUN        | Full Function Node
+| LFN     | Wi-SUN        | Limited Function Node
+| LB-IE  | Wi-SUN        | LFN Broadcast Timing IE
```
indent
